#744: review submit now has separate render and submit methods in public API

garvitverma opened the pull request:
[RETURN_STATUS_DATA] and [PROCESSED_PATHS] are now mandatory outputs to stderr of any nuke hook that gets written for review submit. It's output on stderr to avoid subprocess pipe going into a deadlock. actual exit status of the subprocess determines successful execution from an unsuccessful one.

Commit:
added extra_write_nodes_path_info setting, to store write nodes and their output template, separated out render and submit methods.

Commit:
process the return status from the subprocess, differentiate successful execution and get a reliable stacktrace. Also, get the outputs processed by nuke.

Commit:
make sure the movie is rendered in the hook, before submitting for version creation.

mbrainerd commented on app.py:
```diff
@@ -52,12 +70,62 @@ def render_and_submit(self, template, fields, first_frame, last_frame, sg_publis
 
-    def render_and_submit_version(self, template, fields, first_frame, last_frame, sg_publishes, sg_task,
-                                  comment, thumbnail_path, progress_cb, color_space=None, *args, **kwargs):
+    def render_and_return_paths(self, path, fields, first_frame, last_frame, sg_publishes, sg_task,
```
Just call this `render()`

mbrainerd commented on app.py:
```diff
@@ -71,22 +139,55 @@ def render_and_submit_version(self, template, fields, first_frame, last_frame, s
+        return sg_version
 
-    def render_and_submit_path(self, path, fields, first_frame, last_frame, sg_publishes, sg_task,
+    def render_and_submit_version(self, path, fields, first_frame, last_frame, sg_publishes, sg_task,
```
Please do not change the signature of any previously existing public api methods. We need to maintain both `render_and_submit_version` as well as `render_and_submit_path` as-is.

mbrainerd replied:
Please also change `path` to `path_to_frames` for `render_and_submit_path`

mbrainerd commented on app.py:
```diff
@@ -52,12 +70,62 @@ def render_and_submit(self, template, fields, first_frame, last_frame, sg_publis
+        Create a version entity for the given path.
+
+        :param path:            The path where frames should be found.
+        :param path_to_upload:  The path to create the version entity for.
```
Please change these to `path_to_frames` and `path_to_movie`.

mbrainerd commented on app.py:
```diff
@@ -52,12 +70,62 @@ def render_and_submit(self, template, fields, first_frame, last_frame, sg_publis
+        Render and return the paths that are processed by the nuke hook.
 
-        :param template:        The template defining the path where frames should be found.
+        :param path:            The path where frames should be found.
```
Please change this to `path_to_frames`

mbrainerd commented on app.py:
```diff
@@ -38,6 +38,24 @@ def context_change_allowed(self):
         """
         return True
 
+    def get_resolved_write_nodes(self, fields):
```
`resolve_extra_write_nodes()` would probably be a better name

mbrainerd commented on app.py:
```diff
@@ -129,25 +242,10 @@ def render_and_submit_path(self, path, fields, first_frame, last_frame, sg_publi
-                                      fields.get("name", "Unnamed"),
-                                      color_space, fields)
+        if output_path not in processed_paths:
+            raise Exception("tk-multi-reviewsubmission not configured to render movie!")
```
Please use a more specific Exception

garvitverma replied:
How will specific Exceptions help? Since these modules are can't be imported directly anyways.

Or did you want me to add more details to the exception? Like the processed paths and such?

mbrainerd replied:
The Exception class still provides useful information in the traceback to communicate the developer's intent. Imagine a TD a year from now with no prior knowledge of this tool having to debug an issue. If that was you, what information would you want here?

garvitverma replied:
Makes sense, will add the new Exceptions.

mbrainerd commented on hooks/nuke_batch_render_movie.py:
```diff
@@ -58,7 +59,7 @@ def __create_output_node(path, codec_settings, logger=None):
 
-def render_movie_in_nuke(path, output_path,
+def render_movie_in_nuke(path, output_path, extra_write_node_mapping,
                          width, height,
```
Add the new parameter to the docstring. Also, it should really be added to the end.

garvitverma replied:
It just felt relevant to the path related parameters, should I move it to the end?

mbrainerd commented on python/tk_multi_reviewsubmission/renderer.py:
```diff
@@ -117,22 +114,19 @@ def gather_nuke_render_info(self, path, output_path,
 
         :param path:            The path where frames should be found.
         :param output_path:     The path where the movie should be written to
+        :param extra_write_node_mapping: Mapping of write nodes other than the movie write node with their output paths
```
add: " that should be rendered"

mbrainerd commented on python/tk_multi_reviewsubmission/renderer.py:
```diff
@@ -179,17 +173,35 @@ def render_movie_in_nuke(self, path, output_path,
 
+        processed_paths = thread.get_processed_paths()
+        if not processed_paths:
+            raise Exception("Error in tk-multi-reviewsubmission: No output paths were found after the render!")
```
Please raise a more specific Exception.

mbrainerd reviewed (changes requested):
See inline comments.

Commit:
Updated parameters path to path_to_frames, output_path to path_to_movie.
Restored render_and_submit_path public API method, and render_and_submit_version signature restored.

mbrainerd commented on app.py:
```diff
@@ -187,12 +212,12 @@ def submit_version(self, path, path_to_upload, fields, first_frame, last_frame,
         Main application entry point to be called by other applications / hooks.
 
-        :param path:            The path where frames should be found.
+        :param path_to_frames:            The path where frames should be found.
```
Fix indent

mbrainerd commented on app.py:
```diff
@@ -187,12 +212,12 @@ def submit_version(self, path, path_to_upload, fields, first_frame, last_frame,
 
-    def render_and_submit_version(self, path, fields, first_frame, last_frame, sg_publishes, sg_task,
-                                  comment, thumbnail_path, progress_cb, color_space=None, *args, **kwargs):
+    def render_and_submit_path(self, path_to_frames, fields, first_frame, last_frame, sg_publishes, sg_task, comment,
```
Just for organization sake, please move this so its just below `render_and_submit_version()`

garvitverma replied:
Haha ok :P

mbrainerd reviewed (approved):
Approved pending some minor formatting changes. See inline comments.

Commit:
Added custom Exceptions to renderer.
